logstash-core/src/main/java/org/logstash/common/LsQueueUtils.java

```diff
@@ -49,13 +48,23 @@ public static void addAll(final BlockingQueue<JrubyEventExtLibrary.RubyEvent> qu
         final BlockingQueue<JrubyEventExtLibrary.RubyEvent> queue, final int count, final long nanos
     ) throws InterruptedException {
         int left = count;
+        final long deadline = System.nanoTime() + nanos;
+        long drainTimeout = nanos;
+
         final Collection<JrubyEventExtLibrary.RubyEvent> collection =
             new HashSet<>(4 * count / 3 + 1);
+
         do {
-            final int drained = drain(queue, collection, left, nanos);
+            final int drained = drain(queue, collection, left, drainTimeout);
             if (drained == 0) {
                 break;
             }
+
+            drainTimeout = deadline - System.nanoTime();
+            if (drainTimeout <= 0) {
```

Comment:
@colinsurprenant same as in the other case of this kind of breakout (transitively negative timeouts will lead to  no blocking in a call to `drain` as well), it's just a waste of cycles compared to polling without blocking and breaking out on no `Event`s received.

Reply:
well, if these cycles serve the purpose of being correct WRT to the timeout strategy they are not wasted.